src/server/game/Scripting/ScriptMgr.h

```diff
@@ -710,30 +710,42 @@ class ScriptRegistry
     typedef std::map<uint32, TScript*> ScriptMap;
     typedef typename ScriptMap::iterator ScriptMapIterator;
 
-    typedef std::vector<TScript*> ScriptVector;
+    typedef std::vector<std::pair<TScript*,std::vector<uint16>>> ScriptVector;
     typedef typename ScriptVector::iterator ScriptVectorIterator;
 
+    typedef std::vector<std::vector<TScript*>> EnabledHooksVector;
+    typedef typename EnabledHooksVector::iterator EnabledHooksVectorIterator;
+
     // The actual list of scripts. This will be accessed concurrently, so it must not be modified
     // after server startup.
     static ScriptMap ScriptPointerList;
     // After database load scripts
     static ScriptVector ALScripts;
+    // The list of hook types with the list of enabled scripts for this specific hook.
+    // With this approach, we wouldn't call all available hooks in case if we override just one hook.
+    static EnabledHooksVector EnabledHooks;
 
-    static void AddScript(TScript* const script)
+    static void AddScript(TScript* const script, std::vector<uint16> enabledHooks = std::vector<uint16>())
     {
         ASSERT(script);
 
         if (!_checkMemory(script))
             return;
 
+        if (EnabledHooks.size() == 0)
+            EnabledHooks.resize(script->GetTotalAvailableHooks());
```

Comment:
```suggestion
            EnabledHooks.reserve(script->GetTotalAvailableHooks());
```

Reply:
Here we need resize, since later we are accessing elements like this: `EnabledHooks[v].`

Reply:
Ah, yes, my bad